// @azure/msal-browser/dist/custom-auth-path/custom_auth/core/interaction_client/jit/result/JitActionResult.mjs
/*! @azure/msal-browser v4.23.0 2025-09-17 */
'use strict';
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
// Result type constants
const JIT_GET_AUTH_METHODS_RESULT_TYPE = "JitGetAuthMethodsResult";
const JIT_VERIFICATION_REQUIRED_RESULT_TYPE = "JitVerificationRequiredResult";
const JIT_COMPLETED_RESULT_TYPE = "JitCompletedResult";
function createJitGetAuthMethodsResult(input) {
    return {
        type: JIT_GET_AUTH_METHODS_RESULT_TYPE,
        ...input,
    };
}
function createJitVerificationRequiredResult(input) {
    return {
        type: JIT_VERIFICATION_REQUIRED_RESULT_TYPE,
        ...input,
    };
}
function createJitCompletedResult(input) {
    return {
        type: JIT_COMPLETED_RESULT_TYPE,
        ...input,
    };
}

export { JIT_COMPLETED_RESULT_TYPE, JIT_GET_AUTH_METHODS_RESULT_TYPE, JIT_VERIFICATION_REQUIRED_RESULT_TYPE, createJitCompletedResult, createJitGetAuthMethodsResult, createJitVerificationRequiredResult };
